schema/noaa/createLrgsDb.sql

```diff
@@ -0,0 +1,158 @@
+-- This script creates the tables for the LRGS Database.
+--
+
+--------------------------------------------------------------
+-- Single record LrgsDatabaseVersion table.
+-- Identifies the database and contains version number,
+-- which identifies the table structure.
+--------------------------------------------------------------
+CREATE TABLE lrgs_database
+(
+	db_ver INTEGER NOT NULL,
+	create_time TIMESTAMP,
+	created_by VARCHAR(24),
+	description VARCHAR(1024)
+) ;
+
+INSERT into lrgs_database VALUES(1, NOW(), NULL, NULL);
+
+-- MJM: OpenDCS 6.2 does not support Outage Recovery
+--------------------------------------------------------------
+-- 3 different kinds of gaps. LRGS does recovery in 3 
+-- different ways. The system_outage, domsat_gap and damsnt_outage
+-- will share the same sequence.
+--
+-- system_outage
+--------------------------------------------------------------
+-- CREATE TABLE system_outage
+-- (
+-- 	outage_id INTEGER NOT NULL PRIMARY KEY,
+-- 	begin_time TIMESTAMP,
+-- 	end_time TIMESTAMP,
+-- 	status_code CHAR
+-- );
+
+-- Guarantees no two System_outages have the same ID:
+-- CREATE UNIQUE INDEX System_Outage_IdIdx on system_outage (outage_id) ;
+
+--------------------------------------------------------------
+-- domsat_gap
+--------------------------------------------------------------
+-- CREATE TABLE domsat_gap
+-- (
+-- 	outage_id INTEGER NOT NULL PRIMARY KEY,
+-- 	begin_time TIMESTAMP,
+-- 	begin_seq INTEGER,
+-- 	end_time TIMESTAMP,
+-- 	end_seq INTEGER,
+-- 	status_code CHAR
+-- );
+
+-- Guarantees no two domsat_gap have the same ID:
+-- CREATE UNIQUE INDEX Domsat_Gap_IdIdx on domsat_gap (outage_id) ;
+
+--------------------------------------------------------------
+-- damsnt_outage
+--------------------------------------------------------------
+-- CREATE TABLE damsnt_outage
+-- (
+-- 	outage_id INTEGER NOT NULL PRIMARY KEY,
+-- 	data_source_id INTEGER,
+-- 	begin_time TIMESTAMP,
+-- 	end_time TIMESTAMP,
+-- 	status_code CHAR
+-- );
+
+-- Guarantees no two damsnt_outage have the same ID:
+-- CREATE UNIQUE INDEX Damsnt_Outage_IdIdx on damsnt_outage (outage_id) ;
+
+--------------------------------------------------------------
+-- data_source
+-- Each msg in storage holds a data_source_id.
+-- Data Source uniquely identified by type and name.
+-- DDS & DAMS_NT: name is host/ip-address of server.
+-- Name not used for DOMSAT data source.
+-- Data source record cannot be deleted as long as 
+-- any messages in storage hold it's ID.
+--------------------------------------------------------------
+CREATE TABLE data_source
```

Comment:
based purely on the comment above  I would expect the constraints on the table to be as follows:
```
(
	data_source_id INTEGER NOT NULL primary key,

	-- For multi-LRGS databases, specify which LRGS this is a data source for.
	lrgs_host VARCHAR(64) NOT NULL,
	data_source_name VARCHAR(64) NOT NULL,
	data_source_type VARCHAR(24) NOT NULL,
	UNIQUE(data_source_name,data_source_type) -- perhaps lrgs_host,data_source_name,data_source_type since everything is not null
);
```

Generally when using a surrogate key, you don't combine it it another key, otherwise you end up having to put to column in to a table that references the columns.

Reply:
This table was implemented only for NOAA. At Wallops they have two LRGS servers CDADATA and CDABACKUP. They both share a database used for authentication and statistics gathering. The DB happens to be hosted on Postgres on CDADATA. That's why there's no Oracle version for this script -- nobody else is using it.
The 'lrgs_host' column will either be cdadata or cdabackup. Perhaps if they implement a 3rd LRGS at the station there will be a 3rd possibility. 
So within the day files (not SQL), each message contains a data_source_id referring to one of these records. This capability for a multi-LRGS database was grafted on later. That's the reason for the complex surrogate key.

Reply:
> `That's the reason for the complex surrogate key`

The problem is, a surrogate key should never be complex; it defeats the purpose of a surrogate key which is be devoid of all semantic meaning other that "this single digit/uuid/etc points to this particular row."

(I'm certainly not going to prevent the change from getting in, because as you said for NOAA, but I would be remiss not to point out the longer term maintenance issue of what you're doing.)

Arguments about programming correctness aside, in the above scenario. would it be correct to say that a data source is completely identified by the combination of lrgs_host and data_source_name? thus making (lrgs_host,data_source_name) the natural primary key? 

I would strongly suggest making the table the following: 
```sql
(
	data_source_id INTEGER NOT NULL primary key,

	-- For multi-LRGS databases, specify which LRGS this is a data source for.
	lrgs_host VARCHAR(64) NOT NULL,
	data_source_name VARCHAR(64) NOT NULL,
	data_source_type VARCHAR(24) NOT NULL,
	UNIQUE (lrgs_host,data_source_name,data_source_type) -- this is the "REAL" e.g. natural primary key, but the data_source_id is easier to use externally
);```